### ISSUE TITLE
REST API isnt updating (or I am disconnected after retrieving latest block information)

### ISSUE THREAD

**Issue by lazrfocus** [input-output-hk/jormungandr#853]
**Describe the bug**
- Using Windows 10 WSL Ubuntu terminal
- Running node with command `./jormungandr --config node-config.yaml --genesis-block-hash adbdd5ede31637f6c9bad5c271eec0bc3d0cb9efb86a5b913bb55cba549d0770`
- Config file contains the following:
```
log:
  format: plain
  level: debug
  output: stderr
p2p:
  listen_address: /ip4/0.0.0.0/tcp/3100
  public_address:
  topics_of_interest:
    blocks: high
    messages: high
  trusted_peers:
    - /ip4/3.123.177.192/tcp/3000
    - /ip4/52.57.157.167/tcp/3000
    - /ip4/3.123.155.47/tcp/3000
    - /ip4/3.115.57.216/tcp/3000
    - /ip4/3.112.185.217/tcp/3000
    - /ip4/18.139.40.4/tcp/3000
    - /ip4/18.140.134.230/tcp/3000
rest:
  listen: "127.0.0.1:3101"
storage: /tmp/jormungandr
```

After running initial command, I can get stats of latest block from REST API using jcli or curl:

```
jcli rest v0 node stats get --host http://127.0.0.1:3101/api
---
blockRecvCnt: 0
lastBlockDate: "217.23436"
lastBlockFees: 0
lastBlockHash: 15920ba752bb8df921d8aaddce44ce89501ff94562963dcba4e68dfabf827582
lastBlockHeight: "9907"
lastBlockSum: 0
lastBlockTime: ~
lastBlockTx: 0
txRecvCnt: 0
uptime: 254
```

Repeated runs of this command, however, show the same information.  Block Height and Date do not update.  If I Ctrl-C quit out of the running jormungandr node, and rerun the command to start the node, the REST API will be updated with the latest block information.

```
jcli rest v0 node stats get --host http://127.0.0.1:3101/api
---
blockRecvCnt: 0
lastBlockDate: "217.23806"
lastBlockFees: 0
lastBlockHash: d00a20e04a055a1015978a64d19e08c25e12d36922570d887c89a2c2b22de3e2
lastBlockHeight: "9979"
lastBlockSum: 0
lastBlockTime: ~
lastBlockTx: 0
txRecvCnt: 0
uptime: 5
```

The jormungandr node log prints out the following (is my node disconnecting?)

> **Sep 27 14:07:24.461 DEBG Error sending data: disconnected, scope: global
> Sep 27 14:07:24.462 DEBG dropping I/O source: 0, scope: global**

```
Sep 27 14:07:19.520 DEBG received; frame=Headers { stream_id: StreamId(3), flags: (0x4: END_HEADERS) }, scope: global
Sep 27 14:07:19.521 DEBG received; frame=Headers { stream_id: StreamId(5), flags: (0x4: END_HEADERS) }, scope: global
Sep 27 14:07:19.526 DEBG received; frame=Headers { stream_id: StreamId(7), flags: (0x4: END_HEADERS) }, scope: global
Sep 27 14:07:19.527 DEBG scheduling Read for: 0, scope: global
Sep 27 14:07:19.528 DEBG connected to peer, node_id: 14548023685822520157, peer_addr: 18.139.40.4:3000, task: network
Sep 27 14:07:19.531 DEBG scheduling Read for: 0, scope: global
Sep 27 14:07:19.531 DEBG send; frame=Data { stream_id: StreamId(7) }, scope: global
Sep 27 14:07:24.454 DEBG adding I/O source: 0, scope: global
Sep 27 14:07:24.457 DEBG scheduling Read for: 0, scope: global
Sep 27 14:07:24.458 DEBG scheduling Read for: 0, scope: global
Sep 27 14:07:24.461 DEBG Error sending data: disconnected, scope: global
Sep 27 14:07:24.462 DEBG dropping I/O source: 0, scope: global
Sep 27 14:07:28.932 DEBG scheduling Read for: 0, scope: global
Sep 27 14:07:28.934 DEBG selecting gossips for 7759025128004985606, task: network
Sep 27 14:07:28.935 DEBG sending gossip to node 7759025128004985606, task: network
Sep 27 14:07:28.935 DEBG selecting gossips for 15250992350008249921, task: network
Sep 27 14:07:28.937 DEBG sending gossip to node 15250992350008249921, task: network
Sep 27 14:07:28.938 DEBG selecting gossips for 16729047502099380808, task: network
Sep 27 14:07:28.939 DEBG sending gossip to node 16729047502099380808, task: network
Sep 27 14:07:28.940 DEBG selecting gossips for 14548023685822520157, task: network
Sep 27 14:07:28.941 DEBG sending gossip to node 14548023685822520157, task: network
Sep 27 14:07:28.943 DEBG selecting gossips for 1007204115824491648, task: network
Sep 27 14:07:28.944 DEBG sending gossip to node 1007204115824491648, task: network
Sep 27 14:07:28.946 DEBG selecting gossips for 4485365775303479996, task: network
Sep 27 14:07:28.947 DEBG sending gossip to node 4485365775303479996, task: network
Sep 27 14:07:28.948 DEBG selecting gossips for 15870000297337153516, task: network
Sep 27 14:07:28.949 DEBG sending gossip to node 15870000297337153516, task: network
Sep 27 14:07:28.949 DEBG scheduling Read for: 1, scope: global
Sep 27 14:07:28.950 DEBG scheduling Read for: 1, scope: global
Sep 27 14:07:28.952 DEBG scheduling Read for: 0, scope: global
Sep 27 14:07:28.953 DEBG scheduling Read for: 1, scope: global
Sep 27 14:07:28.955 DEBG send; frame=Data { stream_id: StreamId(7) }, scope: global
Sep 27 14:07:28.956 DEBG send; frame=Data { stream_id: StreamId(7) }, scope: global
```

Mandatory Information
```
bob@DESKTOP-I3HL99H:~/jor$ jcli --version
jcli 0.2.2
bob@DESKTOP-I3HL99H:~/jor$ jormungandr --full-version
jormungandr 0.5.2 (master-e7a3efe7, release, linux [x86_64]) - [rustc 1.38.0 (625451e37 2019-09-23)]
```

**Comment by rinor** [comment]
If your box has only private ip addresses configured, try to put a public ip under public_address:

**Comment by lazrfocus** [comment]
Thank you rinor!  I added the ip of the first trusted peer to the public address.

My config file looks like the following now and the API looks like it is updating as it should!

```
log:
  format: plain
  level: debug
  output: stderr
p2p:
  listen_address: /ip4/0.0.0.0/tcp/3100
  public_address: /ip4/3.123.177.192/tcp/3000
  topics_of_interest:
    blocks: high
    messages: high
  trusted_peers:
    - /ip4/3.123.177.192/tcp/3000
    - /ip4/52.57.157.167/tcp/3000
    - /ip4/3.123.155.47/tcp/3000
    - /ip4/3.115.57.216/tcp/3000
    - /ip4/3.112.185.217/tcp/3000
    - /ip4/18.139.40.4/tcp/3000
    - /ip4/18.140.134.230/tcp/3000
rest:
  listen: "127.0.0.1:3101"
storage: /tmp/jormungandr
```

Should I put a different IP for the public IP? Should I put more in a list, like the trusted peers?

**Comment by lazrfocus** [comment]
After a short amount of time, my node still seems to get disconnected.

```
Sep 27 17:00:42.366 DEBG pulling headers from 1007204115824491648, to: Hash(Blake2b256(0xc1a4eba15b1bfcb4e909aafc1f5943260df15acc3238e7b2d0838dea48407a8c)), from: [Hash(Blake2b256(0x43cbda8dd5a70cc5f2fa5f9b649d5ba8a8eebbdda349f809f9211f7872c4cb60)), Hash(Blake2b256(0x3e18bfa394c134c787e0b894eb553bf3e7b16821016bc14ec143699e41f93ff5)), Hash(Blake2b256(0x06fb8c9c1c138a6e6590923fb0a5caa8c97514bba259ced036db2eab3d5115d6)), Hash(Blake2b256(0xadbdd5ede31637f6c9bad5c271eec0bc3d0cb9efb86a5b913bb55cba549d0770))], task: network
Sep 27 17:00:42.367 DEBG scheduling Read for: 3, scope: global
Sep 27 17:00:42.368 DEBG send; frame=Headers { stream_id: StreamId(229), flags: (0x4: END_HEADERS) }, scope: global
Sep 27 17:00:42.368 DEBG send; frame=Data { stream_id: StreamId(229) }, scope: global
Sep 27 17:00:42.369 DEBG scheduling Read for: 3, scope: global
Sep 27 17:00:42.369 DEBG send; frame=Data { stream_id: StreamId(229), flags: (0x1: END_STREAM) }, scope: global
Sep 27 17:00:42.462 DEBG received; frame=WindowUpdate { stream_id: StreamId(0), size_increment: 25461 }, scope: global
Sep 27 17:00:42.464 DEBG received; frame=WindowUpdate { stream_id: StreamId(7), size_increment: 25160 }, scope: global
Sep 27 17:00:42.464 DEBG scheduling Read for: 3, scope: global
Sep 27 17:00:42.517 DEBG received; frame=Reset { stream_id: StreamId(227), error_code: CANCEL }, scope: global
Sep 27 17:00:42.518 DEBG scheduling Read for: 1, scope: global
Sep 27 17:00:42.519 DEBG client response error: protocol error: stream no longer needed, scope: global
Sep 27 17:00:42.520 DEBG scheduling Read for: 1, scope: global
Sep 27 17:00:42.520 WARN PullBlocksToTip request failed: Error { code: Canceled, source: Status { code: Cancelled, message: "h2 protocol error: protocol error: stream no longer needed" } }, node_id: 4485365775303479996, peer_addr: 3.123.155.47:3000, task: network
Sep 27 17:00:42.532 DEBG received; frame=Reset { stream_id: StreamId(229), error_code: CANCEL }, scope: global
Sep 27 17:00:42.533 DEBG scheduling Read for: 3, scope: global
Sep 27 17:00:42.534 DEBG client response error: protocol error: stream no longer needed, scope: global
Sep 27 17:00:42.534 DEBG scheduling Read for: 3, scope: global
Sep 27 17:00:42.536 WARN PullBlocksToTip request failed: Error { code: Canceled, source: Status { code: Cancelled, message: "h2 protocol error: protocol error: stream no longer needed" } }, node_id: 1007204115824491648, peer_addr: 3.123.177.192:3000, task: network
Sep 27 17:00:42.651 DEBG received; frame=WindowUpdate { stream_id: StreamId(0), size_increment: 25344 }, scope: global
Sep 27 17:00:42.653 DEBG received; frame=WindowUpdate { stream_id: StreamId(7), size_increment: 25184 }, scope: global
Sep 27 17:00:42.653 DEBG scheduling Read for: 28, scope: global
Sep 27 17:00:42.774 DEBG received; frame=WindowUpdate { stream_id: StreamId(0), size_increment: 25320 }, scope: global
Sep 27 17:00:42.774 DEBG received; frame=WindowUpdate { stream_id: StreamId(7), size_increment: 25160 }, scope: global
Sep 27 17:00:42.775 DEBG scheduling Read for: 23, scope: global
Sep 27 17:00:42.790 DEBG received; frame=WindowUpdate { stream_id: StreamId(0), size_incremen
```

**Comment by mjackson001** [comment]
this is happening for me as well.  not a firewall issue, happens when the firewall is completely off.  changing the public IP to my nodes public address doesn't help either.

**Comment by lazrfocus** [comment]
For some reason, the only way I can get updates is if I use one of the trusted peer IPs as the public address.

I have talked to some people in the telegram chat and they said that the public address should be your personal public facing IP address.  

So, I've set it to my public facing IP address, made sure the port was open to the outside world and have my config file setup as the following.  But, I still cant get updates:

```
log:
  format: plain
  level: debug
  output: stderr
p2p:
  listen_address: /ip4/0.0.0.0/tcp/14509
  public_address: /ip4/209.209.238.36/tcp/14509
  topics_of_interest:
    blocks: high
    messages: high
  trusted_peers:
    - /ip4/3.123.177.192/tcp/3000
    - /ip4/52.57.157.167/tcp/3000
    - /ip4/3.123.155.47/tcp/3000
    - /ip4/3.115.57.216/tcp/3000
    - /ip4/3.112.185.217/tcp/3000
    - /ip4/18.139.40.4/tcp/3000
    - /ip4/18.140.134.230/tcp/3000
rest:
  listen: "127.0.0.1:3101"
storage: /mnt/f/jormungandr/testnet
```

If I use a web service to ping my IP and port, I can see the following come through on my node log:

```
Sep 28 18:49:12.782 INFO incoming P2P connection on 10.10.0.8:14509, peer_addr: 46.166.184.225:51516, task: network
Sep 28 18:49:12.784 DEBG scheduling Read for: 0, scope: global
Sep 28 18:49:12.785 DEBG send; frame=Settings { flags: (0x0) }, scope: global
Sep 28 18:49:12.785 DEBG adding I/O source: 0, scope: global
Sep 28 18:49:12.786 DEBG scheduling Write for: 0, scope: global
Sep 28 18:49:12.786 DEBG scheduling Read for: 0, scope: global
Sep 28 18:49:12.787 WARN incoming P2P connection error: Protocol(Error(Io, Custom { kind: ConnectionReset, error: "connection closed unexpectedly" })), peer_addr: 46.166.184.225:51516, task: network
```

I have also set my trusted IP to someone in the telegram chat, and they saw me connect.

I am always able to download to the latest block height when starting my node, but after that, I never get any updates and my node stats look something like this:

```
 jcli rest v0 node stats get -h http://127.0.0.1:3101/api
---
blockRecvCnt: 0
lastBlockDate: "218.32040"
lastBlockFees: 0
lastBlockHash: 98d1064b3b36482859a398ebcbf4cfbc4d25dfb1d6393dc7560ff4583f0af36f
lastBlockHeight: "19000"
lastBlockSum: 0
lastBlockTime: ~
lastBlockTx: 0
txRecvCnt: 0
uptime: 54
```

I am currently stumped on how to proceed correctly.

**Comment by hodlonaut** [comment]
Experiencing the same problem. Everything was working reasonably well with the public_address specified, but then at some point in the arvo my public IP address got changed and from that point onwards nothing is working, even after updating the node config and restarting. 

Sep 29 15:21:32.657 DEBG Error sending data: disconnected, scope: global
Sep 29 15:21:32.657 DEBG dropping I/O source: 0, scope: global

would be good to know what data is being sent, where, what error code it's failing with... and maybe this should be ERR not DEBG level messages?

**Comment by lightiv** [comment]
I am having the same issue here also.  Thanks,

**Comment by lazrfocus** [comment]
Last night someone added only my IP as trusted peer with no blockchain downloaded yet.  My node was online, but not receiving block updates for the reasons stated above.

This person was able to download the entire copy of the blockchain that I had stored on my computer, and then we were both stuck at the same block, both not receiving block updates.

So this points to my network working fine, but something else going wrong.

**Comment by onicrom** [comment]
could be related to this:
https://github.com/input-output-hk/jormungandr/issues/860

**Comment by NicolasDP** [comment]
sorry guys for the late replies. This is a bit difficult to know what is the real problems as it could be for many reasons. The node may not update because it cannot connect to any other nodes.

One of the reasons could be because your cannot join any other peers on the network and this didn't show up. Please update to latest version and let us know if you are still experiencing this issues or seeing new logs

**Comment by mjackson001** [comment]
Same issue but logs have more info:

$ ./jormungandr --config ./node-config.yaml --genesis-block-hash adbdd5ede31637f6c9bad5c271eec0bc3d0cb9efb86a5b913bb55cba549d0770 --log-level=info
Sep 30 17:10:06.343 INFO Starting jormungandr 0.5.4 (HEAD-a9ec5ce, release, linux [x86_64]) - [rustc 1.38.0 (625451e37 2019-09-23)], task: init
Sep 30 17:10:06.343 WARN Node started without path to the stored secret keys (not a stake pool or a BFT leader), task: init
Sep 30 17:10:06.344 INFO storing blockchain in '"/tmp/jormungandr/blocks.sqlite"', task: init
Sep 30 17:10:06.353 INFO fetching block adbdd5ede31637f6c9bad5c271eec0bc3d0cb9efb86a5b913bb55cba549d0770, peer_address: 3.115.57.216:3000, block: adbdd5ede31637f6c9bad5c271eec0bc3d0cb9efb86a5b913bb55cba549d0770, task: init
Sep 30 17:10:06.649 INFO initial bootstrap completed, peer_address: 3.115.57.216:3000, block: adbdd5ede31637f6c9bad5c271eec0bc3d0cb9efb86a5b913bb55cba549d0770, task: init
Sep 30 17:10:06.653 INFO connecting to bootstrap peer 3.123.177.192:3000, peer_addr: 3.123.177.192:3000, task: bootstrap
Sep 30 17:11:21.950 INFO initial bootstrap completed, peer_addr: 3.123.177.192:3000, task: bootstrap
Sep 30 17:11:21.952 INFO starting task, task: client-query
Sep 30 17:11:21.953 INFO starting task, task: network
Sep 30 17:11:21.953 INFO adding P2P Topology module: trusted-peers, task: network
Sep 30 17:11:21.953 INFO start listening and accepting gRPC connections on 0.0.0.0:3000, task: network
Sep 30 17:11:21.954 INFO Starting 1 workers, scope: global
Sep 30 17:11:21.955 INFO Starting server on 127.0.0.1:3101, scope: global
Sep 30 17:11:21.955 INFO connecting to initial gossip peer, peer_addr: 3.112.185.217:3000, task: network
Sep 30 17:11:21.956 INFO connecting to initial gossip peer, peer_addr: 3.115.57.216:3000, task: network
Sep 30 17:11:21.956 INFO connecting to initial gossip peer, peer_addr: 52.57.157.167:3000, task: network
Sep 30 17:11:21.956 INFO connecting to initial gossip peer, peer_addr: 18.139.40.4:3000, task: network
Sep 30 17:11:21.957 INFO connecting to initial gossip peer, peer_addr: 3.123.177.192:3000, task: network
Sep 30 17:11:21.958 INFO connecting to initial gossip peer, peer_addr: 3.123.155.47:3000, task: network
Sep 30 17:11:21.960 INFO connecting to initial gossip peer, peer_addr: 18.140.134.230:3000, task: network
Sep 30 17:11:23.134 INFO incoming P2P connection on 198.58.116.105:3000, peer_addr: 206.189.10.179:51718, task: network
Sep 30 17:11:25.266 INFO incoming P2P connection on 198.58.116.105:3000, peer_addr: 80.57.91.60:51672, task: network
Sep 30 17:11:29.222 INFO incoming P2P connection on 198.58.116.105:3000, peer_addr: 149.248.62.150:50254, task: network
Sep 30 17:11:32.347 INFO peer responded with different node id: 401510887185635334, node_id: 401510887185635334, peer_addr: 138.68.122.238:3000, task: network
Sep 30 17:11:32.476 INFO peer responded with different node id: 4466935019052492952, node_id: 4466935019052492952, peer_addr: 89.162.89.46:8443, task: network
Sep 30 17:11:42.336 INFO peer responded with different node id: 401510887185635334, node_id: 401510887185635334, peer_addr: 138.68.122.238:3000, task: network
Sep 30 17:11:42.568 INFO peer responded with different node id: 4466935019052492952, node_id: 4466935019052492952, peer_addr: 89.162.89.46:8443, task: network
Sep 30 17:11:52.333 INFO peer responded with different node id: 401510887185635334, node_id: 401510887185635334, peer_addr: 138.68.122.238:3000, task: network
Sep 30 17:11:52.523 INFO peer responded with different node id: 4466935019052492952, node_id: 4466935019052492952, peer_addr: 89.162.89.46:8443, task: network

**Comment by hodlonaut** [comment]
> sorry guys for the late replies. This is a bit difficult to know what is the real problems as it could be for many reasons. The node may not update because it cannot connect to any other nodes.
> 
> One of the reasons could be because your cannot join any other peers on the network and this didn't show up. Please update to latest version and let us know if you are still experiencing this issues or seeing new logs

Still experiencing issue with 0.5.4 in my passive node. Adding extra logging would help with troubleshooting, currently just seeing some Error messages at DEBUG level but without details of the error codes, IP addresses involved, etc. There's even no error level log messages showing up about peers  in node output if I for the fun of it turn off my internet connection full stop and start my node... that doesn't sound ideal.

Outside world can see my node just fine as evidenced by incoming connection attempts showing up in node logs while testing open port via https://www.yougetsignal.com/tools/open-ports/ and my public_address / port combination.

**Comment by hodlonaut** [comment]
Quick update - looks like my node finally got a lastBlockHeight update a few minutes ago, so maybe there's hope for me (and others) yet! Hooray. Node stats output went from

blockRecvCnt: 0
lastBlockDate: "221.3023"
lastBlockFees: 0
lastBlockHash: cc661c5cc4153ef94835dff4eee6d77e036a16e5fc619979bcb731c4c12832c6
lastBlockHeight: "35894"
lastBlockSum: 0
lastBlockTime: ~
lastBlockTx: 0
txRecvCnt: 1
uptime: 2844

to

blockRecvCnt: 4561
lastBlockDate: "221.4642"
lastBlockFees: 0
lastBlockHash: 878d0e456b8accb6fd60d1326f084738fa7baaca250dac3467a6f1ef1c2e108b
lastBlockHeight: "35924"
lastBlockSum: 0
lastBlockTime: "2019-10-01T10:28:38+00:00"
lastBlockTx: 0
txRecvCnt: 2
uptime: 3345

**Comment by lazrfocus** [comment]
i am going to close this issue as it is directly related to issue 860[https://github.com/input-output-hk/jormungandr/issues/860](url)
